File: cpp/DecisionTree.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <set>
#include <fstream>
#include <stdlib.h>
#include <string.h>
#define random(x) ( rand() % x )
using namespace std;

struct TreeNode {
	int feature;
	float value;
	string maxClass;
	TreeNode* left = NULL;
	TreeNode* right = NULL;
	/*map<string, int> probas;*/

	TreeNode(int feature, float value) {
		this->feature = feature;
		this->value = value;
	}

	TreeNode(int feature, float value, string maxClass, TreeNode* left,
			TreeNode* right) {
		this->feature = feature;
		this->value = value;
		this->maxClass = maxClass;
		this->left = left;
		this->right = right;
	}
};

class BTree {

public:
	TreeNode* root;

	BTree( void ) {
		root = NULL;
	}

	bool isEmpty() {
		if (root == NULL)
			return true;
		else
			return false;
	}

	void preOrder(TreeNode* node) {
		if (node->left == NULL)
			cout << node->value << "\n";
		else {
			cout << node->value << "\n";
			preOrder(node->left);
			preOrder(node->right);
		}
	}

	int getNumLeafs( TreeNode* node ) {
		int numLeafs = 0;

		if (node->left == NULL)
			numLeafs += 1;

		else {
			numLeafs += getNumLeafs(node->left);
			numLeafs += getNumLeafs(node->right);
		}

		return numLeafs;
	}

	string getClass( vector<float> x, TreeNode* node ) {
		if ( node->left == NULL )
			return node->maxClass;
		else {
			if ( x[node->feature] > node->value )
				return getClass( x, node->right );
			else
				return getClass( x, node->left );
		}
	}
};

float correctNum( vector<string> yTrue, vector<string> yPred ) {
	float ret = 1.0;
	for ( unsigned int i = 0; i < yTrue.size(); i++ ) {
		const char *yt = yTrue[i].c_str();
		const char *yp = yPred[i].c_str();
		if ( strcmp( yt, yp ) == 0 )
			ret += 1.0;
	}
	return ret;
}

float gini(vector<string> y) {
	float ret = 0.0;
	map<string, float> yDict;

	for (unsigned int i = 0; i < y.size(); i++) {
		map<string, float>::iterator iter;
		iter = yDict.find(y[i]);
		if (iter == yDict.end()) {
			yDict[y[i]] = 0;
		}
		yDict[y[i]] += 1;
	}

	map<string, float>::iterator iter = yDict.begin();
	for (; iter != yDict.end(); ++iter) {
		float curr = iter->second;
		ret += (curr / y.size()) * (curr / y.size());
	}

	return 1 - ret;
}

string classify(vector<string> y) {
	map<string, float> yDict;

	for (unsigned int i = 0; i < y.size(); i++) {
		map<string, float>::iterator iter;
		iter = yDict.find(y[i]);
		if (iter == yDict.end()) {
			yDict[y[i]] = 0;
		}
		yDict[y[i]] += 1;
	}

	float max = 0.0;
	string maxId = "none";

	map<string, float>::iterator iter = yDict.begin();
	for (; iter != yDict.end(); ++iter) {
		float curr = iter->second;
		if (curr > max) {
			max = curr;
			maxId = iter->first;
		}
	}

	return maxId;
}

class DecisionTree {

private:

	float tol;
	unsigned int leastSample;
	int maxDepth;
	string merge;
	unsigned int maxFeatures;

public:

	BTree* tree;
	vector<string> targetList;

	DecisionTree(float tol, int leastSample, int maxDepth, string merge,
			int maxFeatures) {
		tree = new BTree();
		this->tol = tol;
		this->leastSample = leastSample;
		this->maxDepth = maxDepth;
		this->merge = merge;
		this->maxFeatures = maxFeatures;
	}

	DecisionTree( void ) {
		tree = new BTree();
		tol = 0.0001;
		leastSample = 1;
		maxDepth = 999;
		merge = "None";
		maxFeatures = 10;
	}

	void binSplitDataSet( vector<vector<float> > &mat0, vector<string> &y0,
			vector<vector<float> > &mat1, vector<string> &y1,
			vector<vector<float> > data, vector<string> y, int feature,
			float value ) {

		mat0.clear();
		y0.clear();
		mat1.clear();
		y1.clear();

		for (unsigned int i = 0; i < data.size(); i++) {
			vector<float> row = data[i];
			if (row[feature] <= value) {
				mat0.push_back(row);
				y0.push_back(y[i]);
			} else {
				mat1.push_back(row);
				y1.push_back(y[i]);
			}
		}
	}

	void chooseBestSplit( vector<vector<float> > data, vector<string> y,
			int &bestIndex, float &bestValue, string &maxClass ) {

		set<string> ySet( y.begin(), y.end() );
		vector<vector<float> > mat0;
		vector<string> y0;
		vector<vector<float> > mat1;
		vector<string> y1;

		if (ySet.size() == 1) {
			bestIndex = -1;
			bestValue = -1;
			maxClass = y[0];
			return;
		}

		int n = data[0].size();
		float S = gini(y);
		float bestS = 1.0;
		bestIndex = 0;
		bestValue = 0.0;
		vector<int> featureList;
		unsigned int visitedFeatures = 0;

		for (int i = 0; i < n; i++)
			featureList.push_back(i);

		while ( visitedFeatures < maxFeatures and featureList.size() > 0 ) {
			srand((int) time(0));
			int randNum = random( featureList.size() );
			int featIndex = featureList[randNum];
			vector<int>::iterator iter = featureList.begin();
			while (iter != featureList.end() && *iter != featIndex)
				iter++;
			if (iter != featureList.end())
				featureList.erase(iter);

			vector<float> column;
			for (unsigned int i = 0; i < data.size(); i++)
				column.push_back(data[i][featIndex]);

			set<float> featSet( column.begin(), column.end() );
			vector<float> featList;
			copy(featSet.begin(), featSet.end(), back_inserter(featList));
			if (featList.size() == 1)
				continue;

			visitedFeatures += 1;
			float lastS = 1.0;

			for (unsigned int i = 0; i < featList.size(); i++) {
				float splitVal = featList[i];
				binSplitDataSet(mat0, y0, mat1, y1, data, y, featIndex, splitVal);

				if (mat0.size() < leastSample || mat1.size() < leastSample)
					continue;

				float r0 = ( (float) y0.size() ) / y.size();
				float r1 = ( (float) y1.size() ) / y.size();
				float newS = r0 * gini(y0) + r1 * gini(y1);

				if ( newS > lastS )
					break;

				if ( newS < bestS ) {
					bestIndex = featIndex;
					bestValue = splitVal;
					bestS = newS;
				}

				lastS = newS;
			}
		}

		maxClass = classify(y);
		if (S - bestS < tol) {
			bestIndex = -1;
			bestValue = -1;
			return;
		}

		binSplitDataSet(mat0, y0, mat1, y1, data, y, bestIndex, bestValue);
		if (mat0.size() < leastSample || mat1.size() < leastSample) {
			bestIndex = -1;
			bestValue = -1;
			return;
		}
	}

	TreeNode* createTree( vector<vector<float> > data, vector<string> y, int depth ) {

		TreeNode* retTree = new TreeNode( -1, -1 );
		if (depth > maxDepth) {
			retTree->maxClass = classify(y);
			return retTree;
		}

		int bestIndex; float bestValue; string maxClass;
		chooseBestSplit( data, y, bestIndex, bestValue, maxClass );
		if ( bestIndex == -1 ) {
			retTree->maxClass = maxClass;
			return retTree;
		}

		vector<vector<float> > left;
		vector<string> yleft;
		vector<vector<float> > right;
		vector<string> yright;
		binSplitDataSet( left, yleft, right, yright, data, y, bestIndex, bestValue );
		retTree->feature = bestIndex;
		retTree->value = bestValue;
		retTree->maxClass = maxClass;
		retTree->left = createTree(left, yleft, depth + 1);
		retTree->right = createTree(right, yright, depth + 1);

		return retTree;
	}

	void train( vector<vector<float> > &data, vector<string> &y ) {

		set<string> targetSet( y.begin(), y.end() );
		copy(targetSet.begin(), targetSet.end(), back_inserter(targetList));
		long length = y.size();
		long cutPoint = length * 9 / 10;
		vector<string> ytrain;
		vector<vector<float> > train;
		vector<string> ytest;
		vector<vector<float> > test;

		/*for ( unsigned int i = 0; i < data.size(); i ++ ) {
			if (i < cutPoint) {
				train.push_back( data[i] );
				ytrain.push_back( y[i] );
			}

			else {
				test.push_back( data[i] );
				ytest.push_back(y[i]);
			}
		}*/

		if (merge == "None") {
			TreeNode* root = createTree(data, y, 0);
			tree->root = root;
		}
	}

	vector<string> predict( vector<vector<float> > data ) {
		vector<string> ret;
		for ( unsigned int i = 0; i < data.size(); i++ ) {
			vector<float> x = data[i];
			string pred = tree->getClass(x, tree->root);
			ret.push_back(pred);
		}
		return ret;
	}

	float score( vector<vector<float> > data, vector<string> y ) {
		vector<string> yPred = predict(data);
		return correctNum( y, yPred ) / (float) y.size();
	}
};

vector<string> splitEx(const string& src, string separate_character) {
	vector<string> strs;
	int separate_characterLen = separate_character.size();
	int lastPosition = 0, index = -1;
	while (-1 != (index = src.find(separate_character, lastPosition))) {
		strs.push_back(src.substr(lastPosition, index - lastPosition));
		lastPosition = index + separate_characterLen;
	}
	string lastString = src.substr(lastPosition); //截取最后一个分隔符后的内容
	if (!lastString.empty())
		strs.push_back(lastString); //如果最后一个分隔符后还有内容就入队
	return strs;
}

void readCSV(vector<vector<float> > &dataTrain, vector<string> &yTrain,
		vector<vector<float> > &dataTest, vector<string> &yTest,
		char* filename) {
	ifstream file(filename);
	string value;
	vector<string> strs;
	int i = 0;

	while (file.good()) {
		getline(file, value, '\n'); // read a string until next comma
		strs = splitEx(value, ",");
		int length = strs.size();

		if (i == 0) {
			i++;
			continue;
		} else if (i < 50000) {
			vector<float> row;
			string y;
			for (int j = 0; j < length; j++) {
				if (j < length - 1)
					row.push_back(atoi(strs[j].c_str()));
				else
					y = strs[j];
			}
			dataTrain.push_back(row);
			yTrain.push_back(y);
		} else {
			vector<float> row;
			string y;
			for (int j = 0; j < length; j++) {
				if (j < length - 1)
					row.push_back(atoi(strs[j].c_str()));
				else
					y = strs[j];
			}
			dataTest.push_back(row);
			yTest.push_back(y);
		}

		i++;
	}
	dataTest.pop_back();
	yTest.pop_back();
}

int main(int argc, char* argv[]) {

	char* filename = "/home/yejiming/desktop/python/ML_Algorithms/dataset.csv";
	vector<vector<float> > dataTrain;
	vector<string> yTrain;
	vector<vector<float> > dataTest;
	vector<string> yTest;
	readCSV(dataTrain, yTrain, dataTest, yTest, filename);

	DecisionTree learner;
	learner.train( dataTrain, yTrain );
	float s = learner.score( dataTest, yTest);
	cout << s;
}

File: java/RandomForest2/DecisionTree.py
package RandomForest2;

import java.io.*;
import java.text.DecimalFormat;
import java.util.*;

import static RandomForest.util.*;


class TreeNode {
	int feature;
	double value;
	String maxClass;
	double[] probas;
	TreeNode left = null;
	TreeNode right = null;
	
	public TreeNode(int feature, double value) {
		this.feature = feature;
		this.value = value;
	}
	
	public TreeNode(int feature, double value, String maxClass, double[] probas, TreeNode left, TreeNode right) {
		this.feature = feature;
		this.value = value;
		this.maxClass = maxClass;
		this.probas = probas;
		this.left = left;
		this.right = right;
	}
}


class BTree {
	
	TreeNode root;
	
	public BTree(TreeNode root) {
		this.root = root;
	}
	
	public BTree() {}

	public void preOrder(TreeNode node) {
		
		if (node.left == null) {
			System.out.print(node.feature);
			System.out.print("   ");
			System.out.print(node.value);
			System.out.print("   ");
			System.out.println(node.maxClass);
		}
		
		else {
			System.out.print(node.feature);
			System.out.print("   ");
			System.out.print(node.value);
			System.out.print("   ");
			System.out.println(node.maxClass);
			preOrder(node.left);
			preOrder(node.right);
		}
		
	}
	
	public int getNumLeafs(TreeNode node) {
		int numLeafs = 0;
		
		if (node.left == null) {
			numLeafs += 1;
		}
		
		else {
			numLeafs += getNumLeafs(node.left);
			numLeafs += getNumLeafs(node.right);
		}
		
		return numLeafs;
	}
	
	public String getClass(double[] x, TreeNode node) {
		
		if (node.left == null) {
			return node.maxClass;
		}
		
		else {
			if (x[node.feature] > node.value)
				return getClass(x, node.right);
			else
				return getClass(x, node.left);
		}
		
	}
	
	public double[] getProba(double[] x, TreeNode node) {
		
		if (node.left == null) {
			return node.probas;
		}
		
		else {
			if (x[node.feature] > node.value)
				return getProba(x, node.right);
			else
				return getProba(x, node.left);
		}
		
	}
}

public class DecisionTree {
	
	double tol = 0.0001;
	int leastSample = 1;
	int maxDepth = 999;
	String merge = "None";
	int maxFeatures = 10;
	List<String> targetList = new ArrayList<String>();
	BTree tree = new BTree();
	
	public DecisionTree(double tol, int leastSample, int maxDepth, String merge, int maxFeatures) {
		this.tol = tol;
		this.leastSample = leastSample;
		this.maxDepth = maxDepth;
		this.merge = merge;
		this.maxFeatures = maxFeatures;
	}
	
	public DecisionTree() {
	}
	
	public List binSplitDataSet(List<DataRow> dataSet, int feature, double value) {
		List ret = new ArrayList();
		List<DataRow> left = new ArrayList<DataRow>();
		List<DataRow> right = new ArrayList<DataRow>();
		
		int size = dataSet.size();
		for (int i = 0; i < size; i++) {
			DataRow row = dataSet.get(i);
			double[] x = row.x;
			if (x[feature] <= value) {
				left.add(row);
			}
			else {
				right.add(row);
			}
		}
		
		ret.add(left); ret.add(right);
		return ret;
	}
	
	public List chooseBestSplit(List<DataRow> dataSet) {
		List ret = new ArrayList();
		int n = dataSet.get(0).x.length;
		int size = dataSet.size();
		
		double S = gini(dataSet);
		if (S == 1.0) {
			ret.add(null); ret.add(null); ret.add(dataSet.get(0).y); ret.add(probaArray(dataSet));
			return ret;
		}
		double bestS = 1.0;
		int bestIndex = 0;
		double bestValue = 0.0;
		List<Integer> featureList = new ArrayList<Integer>();
		int visitedFeatures = 0;
		
		for (int i = 0; i < n; i++)
			featureList.add(i);
		
		while (visitedFeatures < maxFeatures && featureList.size() > 0) {
			Random r = new Random();
			int randNum = r.nextInt(featureList.size());
			int featIndex = featureList.get(randNum);
			featureList.remove(randNum);
			
			List<Double> column = new ArrayList<Double>();
			for (DataRow item : dataSet) { 
				double[] x = item.x;
				column.add(x[featIndex]);
			}
			
			Set<Double> featSet = new HashSet<Double>(column);
			if (featSet.size() == 1)
				continue;
			
			visitedFeatures += 1;
			List<Double> featList = new ArrayList<Double>(featSet);
			Collections.sort(featList);
			Collections.sort(dataSet,new Comparator<DataRow>() {
	            public int compare(DataRow arg0, DataRow arg1) {
	                return ((Double) arg0.x[featIndex]).compareTo((Double) arg1.x[featIndex]);
	            }
	        });
			
			List<Double> column2 = new ArrayList<Double>();
			for (DataRow item : dataSet) { 
				double[] x = item.x;
				column2.add(x[featIndex]);
			}
			int lastIndex = 0;
			
			/*double lastS = (double) 1.0;*/
			
			for (int i = 1; i < featList.size(); i++) {
				double splitVal = featList.get(i);
				int index = column2.indexOf(splitVal);
				
				if (index - lastIndex < leastSample || size - index < leastSample)
					continue;
				
				List<DataRow> ds0 = dataSet.subList(0, index);
				List<DataRow> ds1 = dataSet.subList(index, size - 1);
				double r0 = ((double) ds0.size()) / size;
				double r1 = ((double) ds1.size()) / size;
				double newS = r0 * gini(ds0) + r1 * gini(ds1);
				
				/*if (newS > lastS)
					break;*/
				
				if (newS < bestS) {
					bestIndex = featIndex;
					bestValue = splitVal;
					bestS = newS;
					lastIndex = index;
				}
				
				/*lastS = newS;*/
			}
		}
		
		if (S - bestS < tol) {
			ret.add(null); ret.add(null); ret.add(classify(dataSet)); ret.add(probaArray(dataSet));
			return ret;
		}
		
		List splitted = binSplitDataSet(dataSet, bestIndex, bestValue);
		List<DataRow> ds0 = (List<DataRow>) splitted.get(0);
		List<DataRow> ds1 = (List<DataRow>) splitted.get(1);
		
		if (ds0.size() < leastSample || ds1.size() < leastSample) {
			ret.add(null); ret.add(null); ret.add(classify(dataSet)); ret.add(probaArray(dataSet));
			return ret;
		}
		
		ret.add(bestIndex); ret.add(bestValue); ret.add(classify(dataSet)); ret.add(probaArray(dataSet));
		return ret;
	}
	
	public TreeNode createTree(List<DataRow> dataSet, int depth) {
		TreeNode retTree = new TreeNode(-1, -1);
		
		if (depth > maxDepth) {
			retTree.maxClass = classify(dataSet);
			retTree.probas = probaArray(dataSet);
			return retTree;
		}
		
		List bestSplit = chooseBestSplit(dataSet);
		Object feat = bestSplit.get(0); Object val = bestSplit.get(1); String mc = (String) bestSplit.get(2); double[] probas = (double[]) bestSplit.get(3);
		if (feat == null) {
			retTree.maxClass = mc;
			retTree.probas = probas;
			return retTree;
		}
		
		retTree.feature = (int) feat;
		retTree.value = (double) val;
		retTree.maxClass = mc;
		retTree.probas = probas;
		List splitted = binSplitDataSet(dataSet,  (int) feat, (double) val);
		List<DataRow> left = (List<DataRow>) splitted.get(0);
		List<DataRow> right = (List<DataRow>) splitted.get(1);
		retTree.left = createTree(left, depth + 1);
		retTree.right = createTree(right, depth + 1);
		
		return retTree;
	}
	
	public boolean isLeaf(TreeNode node) {
		return (node.left == null);
	}
	
	private TreeNode prune_pep(TreeNode node, List<DataRow> dataSet) {
		if (isLeaf(node))
			return node;
		
		int numLeafs = tree.getNumLeafs(node);
		long size = dataSet.size();
		List inversed = inverse(dataSet);
		List<double[]> data = (List<double[]>) inversed.get(0);
		List<String> y = (List<String>) inversed.get(1);
		double errorNoMerge = (1 - score(data, y)) * size + 0.5 * numLeafs;
		double varNoMerge = Math.sqrt(errorNoMerge * (1 - errorNoMerge / size));
		List<String> yPred = new ArrayList<String>();
		
		for (int i = 0; i < size; i++) {
			yPred.add(node.maxClass);
		}
		
		double errorMerge = size - correctNum(y, yPred) + 0.5;
		
		if (errorMerge < errorNoMerge + varNoMerge) {
			node.feature = -1; node.value = -1; node.left = null; node.right = null;
			return node;
		}
		
		else {
			List splitted = binSplitDataSet(dataSet, node.feature, node.value);
			List<DataRow> left = (List<DataRow>) splitted.get(0);
			List<DataRow> right = (List<DataRow>) splitted.get(1);
			
			if (!isLeaf(node.left))
				node.left = prune_pep(node.left, left);
			
			if (!isLeaf(node.right))
				node.right = prune_pep(node.right, right);
			
			return node;
		}
	}
	
	public void train(List<double[]> data, List<String> y) {
		Set<String> targetSet = new HashSet<String>(y);
		targetList = new ArrayList<String>(targetSet);
		Collections.sort(targetList);
		
		List<DataRow> dataSet = transform(data, y);
		
		long size = dataSet.size();
		long cutPoint = size * 9 / 10;
		List<DataRow> train = new ArrayList<DataRow>();
		List<DataRow> test = new ArrayList<DataRow>();
		
		
		
		for (int i = 0; i < size; i ++) {
			if (i < cutPoint) {
				train.add(dataSet.get(i));
			}
			
			else {
				test.add(dataSet.get(i));
			}
		}
		
		if (merge == "None") {
			TreeNode root = createTree(dataSet, 0);
			tree = new BTree(root);
		}
		
		else if (merge == "pep" ) {
			TreeNode root = createTree(dataSet, 0);
			tree = new BTree(root);
			root = prune_pep(root, dataSet);
			tree = new BTree(root);
		}
		
		else
			System.out.println("Prune type does not exist!");
	}
	
	public List predict(List<double[]> data) {
		List<String> ret = new ArrayList<String>();
		
		for (double[] x : data) {
			String pred = tree.getClass(x, tree.root);
			ret.add(pred);
		}
		
		return ret;
	}
	
	public double[] probaArray(List<DataRow> dataSet) {
		DecimalFormat df = new DecimalFormat( "0.00000 ");
		double[] ret = new double[targetList.size()];
		int size = dataSet.size();
		
		for (int i = 0; i < ret.length; i ++) 
			ret[i] = (double) 0.0;
		
		for (DataRow item : dataSet) {
			String y = item.y;
			int index = targetList.indexOf(y);
			ret[index] += 1;
		}
		
		for (int i = 0; i < ret.length; i ++) {
			ret[i] /= size;
			ret[i] = Double.parseDouble(df.format(ret[i]));
		}
		
		return ret;
	}
	
	public List<double[]> predictProba(List<double[]> data) {
		List<double[]> ret = new ArrayList<double[]>();
		
		for (double[] x : data) {
			double[] pred = tree.getProba(x, tree.root).clone();
			ret.add(pred);
		}
		
		return ret;
	}
	
	public double score(List<double[]> data, List<String> y) {
		List<String> yPred = new ArrayList<String>();
		yPred = predict(data);
		return correctNum(y, yPred) / ( (double) y.size() );
	}
	
	public static void main(String[] args) throws IOException  {
		String filename = "/home/yejiming/desktop/python/ML_Algorithms/dataset.csv";
		List ret = loadData(filename);
		List<double[]> data = (List<double[]>) ret.get(0);
		List<String> y = (List<String>) ret.get(1);
		List<String> ytrain = new ArrayList<String>();
		List<double[]> train = new ArrayList<double[]>();
		List<String> ytest = new ArrayList<String>();
		List<double[]> test = new ArrayList<double[]>();
		
		for (int i = 0; i < data.size(); i ++) {
			if (i < 50000) {
				train.add(data.get(i));
				ytrain.add(y.get(i));
			}
			
			else {
				test.add(data.get(i));
				ytest.add(y.get(i));
			}
		}
		
		List<DataRow> dataSet = new ArrayList<DataRow>();
		dataSet = transform(train, ytrain);
		System.out.println(dataSet.get(0).x.length);
	}
}
